core/corehttp/gateway_handler.go

```diff
@@ -626,3 +638,42 @@ func webErrorWithCode(w http.ResponseWriter, message string, err error, code int
 func internalWebError(w http.ResponseWriter, err error) {
 	webErrorWithCode(w, "internalWebError", err, http.StatusInternalServerError)
 }
+
+func searchUpTreeFor404(ctx context.Context, acceptHeaders []string, parsedPath coreiface.Path, i *gatewayHandler) (coreiface.Path, error) {
+	filename404, err := preferred404Filename(acceptHeaders)
+	if err != nil {
+		return nil, err
+	}
+
+	pathComponents := strings.SplitAfter(gopath.Dir(parsedPath.String()), "/")
+
+	for idx := len(pathComponents); idx >= 3; idx-- {
+		pretty404 := gopath.Join(append(pathComponents[0:idx], filename404)...)
```

Comment:
I have a hunch this way of walking up the tree has a couple of edge cases, but most of them are fine for now.

Could you add a test for one specific case though? Suppose we request `/foo/bar/`, then `/foo/bar/.ipfs/404.html` should be the first path that's tried. I *think* that it should already be doing that based on `gopath.Dir("/foo/bar/") == "/foo/bar"`, but better make sure.

Reply:
Sorry I meant `/foo/bar/ipfs-404.html` here ^

Reply:
Hey @lgierth, I'm not sure what test you're suggesting; if we request a directory (eg. `/foo/bar/`) then the index page will be returned instead of this new 404 page logic being called. I've updated this PR to meet your other comments, but if you can help me understand your concern here I'll have another look! :)

Reply:
I believe the tests now cover the case you're talking about @lgierth; [this line](https://github.com/ipfs/go-ipfs/pull/4233/files#diff-56f9b8538e5c90481066b55b7feaf780R273) covers that case ☺️